packages/template-retail-react-app/app/components/island/index.jsx

```diff
@@ -0,0 +1,222 @@
+        }
+        child.props?.children && matches.push(...findChildren(child.props.children, componentType))
+    })
+
+    return matches
+}
+
+/**
+ * This component is intended to give developers explicit and fine-granular control over the
+ * hydration behavior of their experiences.
+ * @param {Object} props
+ * @param {ReactNode} [props.children] The child tree
+ * @param {('load' | 'idle' | 'visible' | 'off')} [props.hydrateOn='load'] The island's hydration strategy
+ * @param {{[key: string]: any}} [props.options] Optional strategy-specific options
+ * @param {boolean} [props.clientOnly] Whether to only render the component on the client
+ * @example Default hydration on load
+ *  <Island>
+ *      <MyContent />
+ *  </Island>
+ * @example Hydration on load
+ *  <Island hydrateOn={'load'}>
+ *      <MyContent />
+ *  </Island>
+ * @example Hydration on idle
+ *  <Island hydrateOn={'idle'}>
+ *      <MyContent />
+ *  </Island>
+ * @example Hydration on idle with custom `requestIdleCallback` options
+ *  <Island hydrateOn={'idle'} options={{timeout: 500}}>
+ *      <MyContent />
+ *  </Island>
+ * @example Hydration on visible (default `IntersectionObserver` options = `{rootMargin: '250px'}`)
+ *  <Island hydrateOn={'visible'}>
+ *      <MyContent />
+ *  </Island>
+ * @example Hydration on visible with custom `IntersectionObserver` options
+ *  <Island hydrateOn={'visible'} options={{rootMargin: '100px'}}>
+ *      <MyContent />
+ *  </Island>
+ * @example Hydration turned off
+ *  <Island hydrateOn={'off'}>
+ *      <MyContent />
+ *  </Island>
+ * @see {@link https://docs.astro.build/en/concepts/islands}
+ * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback}
+ * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver}
+ */
+function Island(props) {
+    const {children, hydrateOn = 'load', options, clientOnly = false} = props
+    const ssr = isServer()
+    const [hydrated, setHydrated] = useState(ssr) // Ensure SSR immediately returns the generated HTML
+    const context = useIslandContext()
+    const data = context ? context.data : new Map()
+    const ref = useRef(null)
+    const hydrate = () => {
+        setHydrated(true)
+    }
+
+    const useIsomorphicLayoutEffect = ssr ? useEffect : useLayoutEffect
+    useIsomorphicLayoutEffect(() => {
+        if (
+            !ssr &&
+            !hydrated &&
+            hydrateOn !== 'off' &&
+            (!context || context?.hydrated) &&
+            ref.current &&
+            !ref.current?.childElementCount &&
+            !clientOnly
+        ) {
+            // No actual SSR Content or rendering after CSR navigation --> hydrate
+            hydrate()
+        }
+    })
+
+    useEffect(() => {
+        if (
+            hydrated ||
+            hydrateOn === 'off' ||
+            (context && !context?.hydrated) ||
+            !ref.current ||
+            (!ref.current?.firstElementChild && !ssr && !clientOnly)
```

Comment:
Can we be consistent with line 100, so that we are both checking for element count or first element?